lib/pengine/rules.c

```diff
@@ -1105,6 +839,481 @@ pe_unpack_versioned_parameters(xmlNode *versioned_params, const char *ra_version
+
+    duration_spec = first_named_child(expr, "duration");
+    date_spec = first_named_child(expr, "date_spec");
+
+    value = crm_element_value(expr, "start");
+    if (value != NULL) {
+        start = crm_time_new(value);
+    }
+    value = crm_element_value(expr, "end");
+    if (value != NULL) {
+        end = crm_time_new(value);
+    }
+
+    if (start != NULL && end == NULL && duration_spec != NULL) {
+        end = pe_parse_xml_duration(start, duration_spec);
+    }
+
+    if ((op == NULL) || safe_str_eq(op, "in_range")) {
+        if ((start == NULL) && (end == NULL)) {
+            // in_range requires at least one of start or end
+        } else if ((start != NULL) && (crm_time_compare(rule_data->now, start) < 0)) {
+            rc = pcmk_rc_before_range;
+            crm_time_set_if_earlier(next_change, start);
+        } else if ((end != NULL) && (crm_time_compare(rule_data->now, end) > 0)) {
+            rc = pcmk_rc_after_range;
+        } else {
+            rc = pcmk_rc_within_range;
+            if (end && next_change) {
+                // Evaluation doesn't change until second after end
+                crm_time_add_seconds(end, 1);
+                crm_time_set_if_earlier(next_change, end);
+            }
+        }
+
+    } else if (safe_str_eq(op, "date_spec")) {
+        rc = pe_cron_range_satisfied(rule_data->now, date_spec);
+        // @TODO set next_change appropriately
+
+    } else if (safe_str_eq(op, "gt")) {
+        if (start == NULL) {
+            // gt requires start
+        } else if (crm_time_compare(rule_data->now, start) > 0) {
+            rc = pcmk_rc_within_range;
+        } else {
+            rc = pcmk_rc_before_range;
+
+            // Evaluation doesn't change until second after start
+            crm_time_add_seconds(start, 1);
+            crm_time_set_if_earlier(next_change, start);
+        }
+
+    } else if (safe_str_eq(op, "lt")) {
+        if (end == NULL) {
+            // lt requires end
+        } else if (crm_time_compare(rule_data->now, end) < 0) {
+            rc = pcmk_rc_within_range;
+            crm_time_set_if_earlier(next_change, end);
+        } else {
+            rc = pcmk_rc_after_range;
+        }
+    }
+
+    crm_time_free(start);
+    crm_time_free(end);
+    return rc;
+}
+
+gboolean
+pe__eval_op_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data) {
+    const char *name = crm_element_value(expr, XML_NVPAIR_ATTR_NAME);
+    const char *interval_s = crm_element_value(expr, XML_LRM_ATTR_INTERVAL);
+    guint interval;
+
+    crm_trace("Testing op_defaults expression: %s", ID(expr));
+
+    if (rule_data->op_data == NULL) {
+        crm_trace("No operations data provided");
+        return FALSE;
+    }
+
+    interval = crm_parse_interval_spec(interval_s);
+    if (interval == 0 && errno != 0) {
+        crm_trace("Could not parse interval: %s", interval_s);
+        return FALSE;
+    }
+
+    if (interval != 0 && interval != rule_data->op_data->interval) {
```

Comment:
I would use interval_s != NULL instead of interval != 0 ... the configuration could explicitly specify 0 to match only probes

Reply:
Done.